tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,711 @@
+
+	headChunksFormatV1 = 1
+	inBufferShards     = 16
+	writeBufferSize    = 2 * 1024 * 1024
+)
+
+var (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
+	// Assuming a general scrape interval of 15s, a chunk with 120 samples would
+	// be cut every 30m, so anything <30m will cause lots of empty files. And keeping
+	// it exactly 30m also has a chance of having empty files as its near that border.
+	// Hence keeping it a little more than 30m, i.e. 40m.
+	DefaultHeadChunkSegmentTime = 40 * int64(time.Minute/time.Millisecond)
+
+	// ErrHeadReadWriterClosed returned by any method indicates
+	// that the HeadReadWriter was closed.
+	ErrHeadReadWriterClosed = errors.New("HeadReadWriter closed")
+)
+
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
+	// Writer.
+	dirFile *os.File
+	curFile *os.File // Segment file being written to.
+
+	curFileMint     int64 // In milliseconds.
+	curFileMaxt     int64 // In milliseconds.
+	curFileSequence int
+	n               int64 // Bytes written in current segment.
+	segmentTime     int64 // Duration in milliseconds.
+
+	buf                 [8]byte
+	wbuf                *bufio.Writer
+	wbufLock            sync.RWMutex
+	inBufferChunks      [inBufferShards]map[uint64]chunkenc.Chunk
+	inBufferChunksLocks [inBufferShards]sync.RWMutex
+
+	// Reader.
+	bs map[int]ByteSlice
+	cs map[int]io.Closer // Closers for resources behind the byte slices.
+
+	size  int64 // The total size of bytes in the reader.
+	bsMtx sync.RWMutex
+	pool  chunkenc.Pool
+
+	quit chan struct{}
+}
+
+const (
+	// HeadSegmentHeaderSize is the total size of the header for the segment file.
+	HeadSegmentHeaderSize = SegmentHeaderSize + 16
+	// HeaderMintOffset is the offset where the first byte of MinT for segment file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for segment file exists.
+	HeaderMaxtOffset = HeaderMintOffset + 8
+	// MaxSegmentSize is the max size of a segment file.
+	// Setting size to the max int32 as setting it to max int 64 crashes 64 systems too.
+	MaxSegmentSize = math.MaxInt32
+)
+
+// NewHeadReadWriter returns a new writer against the given directory
+// using the default segment time.
+func NewHeadReadWriter(dir string, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	return newHeadReadWriter(dir, DefaultHeadChunkSegmentTime, pool)
+}
+
+func newHeadReadWriter(dir string, segmentTime int64, pool chunkenc.Pool) (*HeadReadWriter, error) {
+	if segmentTime <= 0 {
+		segmentTime = DefaultHeadChunkSegmentTime
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	hrw := &HeadReadWriter{
+		dirFile:     dirFile,
+		n:           0,
+		segmentTime: segmentTime,
+		pool:        pool,
+		quit:        make(chan struct{}),
```

Comment:
Boolean also works. Just a habit to use a channel.

Reply:
And usually in this case I believe it should be atomic load and store else it would be termed as a race in some cases. I don't see boolean load and store in `atomic` module. Any suggestions?

Reply:
I don't see where this can cause a race, this api is never to be used concurently right? like the rest of the TSDB api.

Reply:
`WriteChunk` and `Chunk` can be called concurrently, it is safe. But I see the closing part is protected by mutex. I will change it to a boolean.